rotary_utils/external.py

```diff
@@ -0,0 +1,273 @@
+                                         text=True, log=log).stdout
+
+        # Special parsing is needed for samtools output.
+        if dependency_name == 'samtools':
+            dependency_version = stdout.splitlines()[0].split(' ')[1]
+        else:
+            dependency_version = stdout.splitlines()[0]
+    else:
+        dependency_version = 'unknown'
+    return dependency_version
+
+
+def run_pipeline_subcommand(command_args, stdin=None, stdout=None, stderr=None, check=True, text=None, log=True):
+    """
+    Wrapper function for running subcommands.
+
+    :param command_args: The command line arguments of the subcommand (e.g., ['samtools', '-h'])
+    :param stdin: A subprocess.PIPE or None if stdin is not to be used.
+    :param stdout: Where to send stdout or None if stdout is not to be used.
+    :param stderr: Where to send stderr or None if stderr is not to be used.
+    :param check: Cause a runtime error if the subcommand fails.
+    :param text: If True, open outputs in text mode (rather than binary)
+    :param log: If True, write the shell command to logger in debug mode
+    :return: The output of the subcommand.
+    """
+
+    if log is True:
+        logger.debug(shlex.join(command_args))
+    if stdin:
+        stdin = stdin.stdout
+
+    return subprocess.run(command_args, check=check, input=stdin, stdout=stdout, stderr=stderr, text=text)
+
+
+def map_long_reads(contig_filepath: str, long_read_filepath: str, output_bam_filepath: str, log_filepath: str,
+                   append_log: bool = True, threads: int = 1, threads_mem_mb: float = 1):
+    """
+    Maps long reads (via minimap2) to contigs and sorts/indexes the resulting BAM file. output_bam_filepath and
+    log_filepath are saved to disk.
+
+    :param contig_filepath: path to the FastA file containing the reference contigs
+    :param long_read_filepath: path to the FastQ file containing long reads to map (compressed is OK)
+    :param output_bam_filepath: path to the BAM file to be saved
+    :param log_filepath: path to the log file to be saved
+    :param append_log: whether the log should append to an existing file (True) or overwrite an existing file (False);
+                       this setting is only relevant if the log file at log_filepath already exists
+    :param threads: number of threads to use for read mapping
+    :param threads_mem_mb: memory in MB per thread (to use for samtools); must be an integer
+    """
+
+    write_mode = set_write_mode(append_log)
+
+    with open(log_filepath, write_mode) as logfile_handle:
+        with open(output_bam_filepath, 'w') as bam_handle:
+            # TODO - add support for different flags like -ax for pacbio
+            minimap_args = ['minimap2', '-t', str(threads), '-ax', 'map-ont', contig_filepath, long_read_filepath]
+            minimap = run_pipeline_subcommand(command_args=minimap_args, stdout=subprocess.PIPE, stderr=logfile_handle)
```

Comment:
@jmtsuji Here, we might be able to implement a function that calls a chain of subcommands piped into each other.

Also, note that there may be more sophisticated libraries for performing this kind of work, like https://pypi.org/project/sh/, that remove some of the boilerplate. I was initially going to use `sh` for rotary, but there were some features that were missing so I did not use it. It still might be useful for your code.

Reply:
@jmtsuji As you can see, moving all the subcommand code to `external.py' makes switching to a new library like `sh` much easier because all the code that uses subcommands is in one place.

Reply:
@LeeBergstrand Interesting idea! At the same time, the current code already seems fairly compact and workable. To prioritize other code dev and refactoring in this rapid development phase, would you be OK if I convert this to an issue (enhancement) for now? We could then follow up on this in the future, especially if more externals are added. Or do you think the current code could run into problems that we should address at this stage? Thanks!

Reply:
@jmtsuji Fair enough, added as issue https://github.com/rotary-genomics/rotary-utils/issues/7